test/test-udp-connect6.c

```diff
@@ -0,0 +1,197 @@
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define CHECK_HANDLE(handle) \
+  ASSERT((uv_udp_t*)(handle) == &server || (uv_udp_t*)(handle) == &client)
+
+static uv_udp_t server;
+static uv_udp_t client;
+static uv_buf_t buf;
+static struct sockaddr_in6 lo_addr;
+
+static int cl_send_cb_called;
+static int sv_recv_cb_called;
+
+static int close_cb_called;
+
+
+static void alloc_cb(uv_handle_t* handle,
+                     size_t suggested_size,
+                     uv_buf_t* buf) {
+  static char slab[65536];
+  CHECK_HANDLE(handle);
+  ASSERT(suggested_size <= sizeof(slab));
+  buf->base = slab;
+  buf->len = sizeof(slab);
+}
+
+
+static void close_cb(uv_handle_t* handle) {
+  CHECK_HANDLE(handle);
+  ASSERT(uv_is_closing(handle));
+  close_cb_called++;
+}
+
+
+static void cl_send_cb(uv_udp_send_t* req, int status) {
+  int r;
+
+  ASSERT_NOT_NULL(req);
+  ASSERT(status == 0);
+  CHECK_HANDLE(req->handle);
+  if (++cl_send_cb_called == 1) {
+    uv_udp_connect(&client, NULL);
+    r = uv_udp_send(req, &client, &buf, 1, NULL, cl_send_cb);
+    ASSERT(r == UV_EDESTADDRREQ);
+    r = uv_udp_send(req,
+                    &client,
+                    &buf,
+                    1,
+                    (const struct sockaddr*) &lo_addr,
+                    cl_send_cb);
+    ASSERT(r == 0);
+  }
+
+}
+
+
+static void sv_recv_cb(uv_udp_t* handle,
+                       ssize_t nread,
+                       const uv_buf_t* rcvbuf,
+                       const struct sockaddr* addr,
+                       unsigned flags) {
+  if (nread > 0) {
+    ASSERT(nread == 4);
+    ASSERT_NOT_NULL(addr);
+    ASSERT(memcmp("EXIT", rcvbuf->base, nread) == 0);
+    if (++sv_recv_cb_called == 4) {
+      uv_close((uv_handle_t*) &server, close_cb);
+      uv_close((uv_handle_t*) &client, close_cb);
+    }
+  }
+}
+
+
+TEST_IMPL(udp_connect6) {
```

Comment:
Thanks for adding this test. It may uncover the same issue in the BSD's, who knows.

Reply:
What about use macro `BSD` instead `__APPLE__`?

Reply:
Let's see first how https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2203/ goes

Reply:
And from the `freebsd-src`, seems freebsd latest support disconnect a udp socket: https://github.com/freebsd/freebsd-src/blob/main/sys/kern/uipc_socket.c#L1374

Reply:
> Let's see first how https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2203/ goes

I have checked the freebsd test result `-22` is EINVAL, so let's change the macro test from `__APPLE__` to `BSD` and test again?

Reply:
Yes please, but for the moment, I'd just check `__FreeBSD__` and `__FreeBSD_kernel__` as sometimes the BSD differ. Let's wait for bug reports before adding the rest of them. (Unless you have ways to check on the other BSD's, of course)

Reply:
Done.

Reply:
Maybe, we can use github vm action to test `openbsd` and `netbsd`:
https://github.com/vmactions

Reply:
Hi, I do some dig into bsd like OS kernel source code, such freebsd,openbsd,netbsd,DragonFlyBSD, both of them have It seems same `connect` step for udp socket:
1. Check if udp socket connect, invoke `sodisconnect` to disconnect it
2. IF disconnect succed (ret=0), will always continue try to connect with the new sockaddr
So maybe this issue is the common behavior of all BSDs，and maybe this PR logic could be change to simply ignore `EINVAL` and `ENOSUPPORT` for them
a. EINVAL: addrlen mismatch
b. ENOSUPPORT: family mismatch